File: wrappers/python/src/amplifier_agent_client/display.py
"""Display adapter — onEvent push callback + sub-agent event filtering.

Pattern reference: design §4.5 sub-agent leak control.

apply_display_filter(*, subagent_events='all') returns a keep predicate
    Callable[[DisplayEvent], bool].

If subagent_events == 'all': keep everything.
If subagent_events == 'none': suppress events whose parent_turn_id is set.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from amplifier_agent_client.session import DisplayEvent

#: Allowed sub-agent filter mode values.
SubagentMode = Literal["all", "none"]


def apply_display_filter(
    *,
    subagent_events: SubagentMode = "all",
) -> Callable[[DisplayEvent], bool]:
    """Build a keep-predicate from the subagent_events mode.

    Args:
        subagent_events: 'all' (default) keeps every event;
                         'none' suppresses events with parent_turn_id set.

    Returns:
        A predicate ``(ev: DisplayEvent) -> bool`` that returns True iff
        the event should be delivered to the consumer.
    """
    if subagent_events == "all":
        return lambda _ev: True

    # subagent_events == 'none': suppress events carrying parent_turn_id.
    return lambda ev: ev.parent_turn_id is None

File: wrappers/typescript/src/display.ts
/**
 * Display adapter — onEvent push callback + sub-agent event filtering.
 *
 * Pattern reference: design §4.5 sub-agent leak control.
 *
 * `applyDisplayFilter(adapter)` returns a predicate `(ev: DisplayEvent) => boolean`.
 * Mode = adapter.subagentEvents ?? 'all'.
 * If mode === 'all': return true for every event.
 * If mode === 'none': return true only for events whose parentTurnId is null/undefined.
 */

import type { DisplayEvent } from "./session.js";

/** Sub-agent filter mode: 'all' keeps everything, 'none' drops sub-agent events. */
export type SubagentMode = "all" | "none";

/**
 * Display adapter supplied by the host.
 *
 * - `onEvent`: optional push callback invoked for every kept event (pull via iterator
 *   AND push via callback see the same filtered stream).
 * - `subagentEvents`: 'all' (default) keeps sub-agent events; 'none' suppresses them.
 *   Sub-agent events are identified by the presence of `parentTurnId` in the event.
 */
export interface DisplayAdapter {
  onEvent?: (event: DisplayEvent) => void;
  subagentEvents?: SubagentMode;
}

/**
 * Build a keep-predicate from a DisplayAdapter.
 *
 * @param adapter - Host-supplied DisplayAdapter (may have no properties set).
 * @returns A predicate `(ev: DisplayEvent) => boolean` that returns true iff
 *          the event should be delivered to the consumer.
 */
export function applyDisplayFilter(
  adapter: DisplayAdapter,
): (ev: DisplayEvent) => boolean {
  const mode: SubagentMode = adapter.subagentEvents ?? "all";

  if (mode === "all") {
    return (_ev: DisplayEvent): boolean => true;
  }

  // mode === 'none': suppress events that carry parentTurnId (sub-agent events).
  return (ev: DisplayEvent): boolean => ev.parentTurnId == null;
}
